lang/cs/Org.Apache.REEF.IMRU/OnREEF/CheckpointHandler/IMRUCheckpointHandler.cs

```diff
@@ -0,0 +1,128 @@
+
+        private readonly IFileSystem _fileSystem;
+        private readonly string _checkpointFilePath;
+        private const string StateDir = "StateDir";
+        private const string StateFile = "StateFile";
+        private const string FlagFile = "FlagFile";
+        public const string StateFileExt = ".bin";
+        public const string FlagFileExt = ".txt";
+
+        /// <summary>
+        /// It is for storing and retrieving checkpoint data.
+        /// </summary>
+        /// <param name="checkpointFilePath">The file path where the checkpoint data will be stored.</param>
+        /// <param name="fileSystem">File system to load/upload checkpoint data</param>
+        [Inject]
+        private IMRUCheckpointHandler(
+            [Parameter(typeof(CheckpointFilePath))] string checkpointFilePath,
+            IFileSystem fileSystem)
+        {
+            _fileSystem = fileSystem;
+            _checkpointFilePath = checkpointFilePath;
+
+            Logger.Log(Level.Info, "State file path: {0}", checkpointFilePath);
+        }
+
+        /// <summary>
+        /// Save serialized checkpoint data to remote checkpoint file.
+        /// </summary>
+        /// <param name="taskState"></param>
+        /// <param name="codec"></param>
+        public void Persistent(ITaskState taskState, ICodec<ITaskState> codec)
+        {
+            var localStateFile = Path.GetTempPath() + Guid.NewGuid().ToString("N").Substring(0, 4);
+            var localFlagfile = Path.GetTempPath() + Guid.NewGuid().ToString("N").Substring(0, 4);
+
+            string tick = DateTime.Now.Ticks.ToString();
+            string stateFileDir = Path.Combine(_checkpointFilePath, StateDir + tick);
+            string remoteStateFileName = Path.Combine(stateFileDir, StateFile + tick + StateFileExt);
+            string remoteFlagFileName = Path.Combine(stateFileDir, FlagFile + tick + FlagFileExt);
+
+            var stateFileUri = _fileSystem.CreateUriForPath(remoteStateFileName);
+            var flagFileUri = _fileSystem.CreateUriForPath(remoteFlagFileName);
+
+            var data = codec.Encode(taskState);
+            File.WriteAllBytes(localStateFile, data);
+            File.WriteAllText(localFlagfile, remoteStateFileName);
+
+            _fileSystem.CopyFromLocal(localStateFile, stateFileUri);
+            _fileSystem.CopyFromLocal(localFlagfile, flagFileUri);
+
+            File.Delete(localStateFile);
+            File.Delete(localFlagfile);
+        }
+
+        /// <summary>
+        /// Read checkpoint data and deserialize it into ITaskState object.
+        /// </summary>
+        /// <param name="codec"></param>
+        /// <returns></returns>
+        public ITaskState Restore(ICodec<ITaskState> codec)
+        {
+            if (!string.IsNullOrEmpty(_checkpointFilePath))
+            {
+                var files = _fileSystem.GetChildren(_fileSystem.CreateUriForPath(_checkpointFilePath));
```

Comment:
The URI should have been created in the constructor.

Reply:
Yes, it can be. Persist can be called many times and restore is called only once for a single recovery. So not much to optimize.